packages/backend-storage/src/storage_access_arbiter.ts

```diff
@@ -0,0 +1,145 @@
+import {
+  ResourceAccessAcceptor,
+  SsmEnvironmentEntry,
+} from '@aws-amplify/plugin-types';
+import { StorageAction, StoragePath } from './types.js';
+import { StorageAccessPolicyFactory } from './storage_access_policy_factory.js';
+
+// some types internal to this file to improve readability
+type AcceptorToken = string;
+type SetDenyByDefault = (storagePath: StoragePath) => void;
+
+/**
+ * Internal collaborating class for building up mappings of the actions and s3 prefix resource scopes that should be granted to each ResourceAccessAcceptors.
+ * This class maintains two different data structures on top of the same underlying values.
+ * The first is a "forward map" of the allow/deny rules for each access acceptor
+ * The second if a "reverse map" of a "DenyByDefault" callback list for each s3 prefix that has been seen.
+ * This callback is used during a final pass over all of the rules to update the forward map with deny rules on any paths that do not have explicit allow rules
+ */
+export class StorageAccessArbiter {
```

Comment:
I'm open to name suggestions for this class. This name isn't super great but everything else I could think of quickly turned into a Java class name meme

Reply:
After reading internals - perhaps `StorageAccessBuilder` ?

Reply:
`StorageAccessBuilder` is already the name of the type we expose to customers for building their auth rules. We could rename that. However, the more I think about it, the more I want to roll both this class and the "orchestrator" class into a single "StorageAccessExecutor" or "StorageAccessEvaluator"